src/main/kotlin/org/rust/ide/typing/paste/JsonStructParser.kt

```diff
@@ -0,0 +1,173 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.typing.paste
+
+import com.fasterxml.jackson.core.JsonFactory
+import com.fasterxml.jackson.core.JsonParser
+import com.fasterxml.jackson.core.JsonToken
+import java.io.IOException
+
+sealed class DataType {
+    object Integer : DataType()
+    object Float : DataType()
+    object Boolean : DataType()
+    object String : DataType()
+    object Unknown : DataType()
+    data class StructRef(val struct: Struct) : DataType()
+    data class Array(val type: DataType) : DataType()
+    data class Nullable(val type: DataType) : DataType()
+}
+
+class Struct(val fields: LinkedHashMap<String, DataType>) {
+    // Consider structs with the same field names and types to be the same
+    // The fields are sorted before comparison, so two struct with the same field types and names, but
+    // with a different field order, are considered to be the same struct.
+    override fun equals(other: Any?): Boolean {
```

Comment:
Can't we make `Struct` to be `data class` and use generated `equals` and `hashCode`?

Reply:
I thought that it would be better to compare them regardless of the field order.
I want to:
1) Know about the field order, so that `{"a": 1, "b": true}` will generate `struct Struct { a: i64, b: bool }`, but
2) De-duplicate structs regardless of the input order, so that
```json
[{"a": 1, "b": true}, {b: false, "a": 2}]
```
will create only a single struct.

Or do you think that we should create two structs in that case?

Reply:
> I thought that it would be better to compare them regardless of the field order. I want to:
> 
> 1. Know about the field order, so that `{"a": 1, "b": true}` will generate `struct Struct { a: i64, b: bool }`, but
> 2. De-duplicate structs regardless of the input order, so that
> 
> ```json
> [{"a": 1, "b": true}, {"b": false, "a": 2}]
> ```
> 
> will create only a single struct.

You are right, it should work like this

But looks like `LinkedHashMap::equals` compares regardless of order, so it suits our needs